libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp

```diff
@@ -0,0 +1,451 @@
+    ret_scha63t &= RegisterRead(SCHA63T_DUE, TEMP, rsp_due_temper);
+    // RATE_X Cmd Send + ACCL_Z Response Receive
+    ret_scha63t &= RegisterRead(SCHA63T_UNO, RATE_XZ, rsp_rate_x);
+    // TEMPER Cmd Send + TEMPRE Response Receive
+    ret_scha63t &= RegisterRead(SCHA63T_UNO, TEMP, rsp_rate_x);
+    // TEMPER Cmd Send + TEMPRE Response Receive
+    ret_scha63t &= RegisterRead(SCHA63T_UNO, TEMP, rsp_uno_temper);
+
+    // response data address check
+    if (((rsp_rate_x[0] & 0x7C) >> 2) == RATE_XZ) {
+        gyro_x = combine(rsp_rate_x[1], rsp_rate_x[2]);
+    } else {
+        ret_scha63t &= false;
+    }
+    if (((rsp_rate_y[0] & 0x7C) >> 2) == RATE_Y) {
+        gyro_y = combine(rsp_rate_y[1], rsp_rate_y[2]);
+    } else {
+        ret_scha63t &= false;
+    }
+    if (((rsp_rate_z[0] & 0x7C) >> 2) == RATE_XZ) {
+        gyro_z = combine(rsp_rate_z[1], rsp_rate_z[2]);
+    } else {
+        ret_scha63t &= false;
+    }
+    if (((rsp_uno_temper[0] & 0x7C) >> 2) == TEMP) {
+        uno_temp = combine(rsp_uno_temper[1], rsp_uno_temper[2]);
+    } else {
+        ret_scha63t &= false;
+    }
+    if (((rsp_due_temper[0] & 0x7C) >> 2) == TEMP) {
+        due_temp = combine(rsp_due_temper[1], rsp_due_temper[2]);
+    } else {
+        ret_scha63t &= false;
+    }
+    set_temperature(gyro_instance, (uno_temp + due_temp) / 2);
+
+    // change coordinate system from left hand too right hand
+    gyro_z = (gyro_z == INT16_MIN) ? INT16_MAX : -gyro_z;
+
+    Vector3f gyro(gyro_x, gyro_y, gyro_z);
+    gyro *= radians(1.f / 80.f);
+
+    _rotate_and_correct_gyro(gyro_instance, gyro);
+    _notify_new_gyro_raw_sample(gyro_instance, gyro);
+
+    AP_HAL::Device::checkreg reg;
+    if (!dev_due->check_next_register(reg)) {
+        log_register_change(dev_due->get_bus_id(), reg);
+        _inc_gyro_error_count(gyro_instance);
+    }
+}
+
+void AP_InertialSensor_SCHA63T::set_temperature(uint8_t instance, uint16_t temper)
+{
+    float temperature = 25.0f + ( temper / 30 );
+    float temp_degc = (0.5f * temperature) + 23.0f;
+    _publish_temperature(instance, temp_degc);
+}
+
+bool AP_InertialSensor_SCHA63T::update()
+{
+    update_accel(accel_instance);
+    update_gyro(gyro_instance);
+    return true;
+}
+
+bool AP_InertialSensor_SCHA63T::RegisterRead(int uno_due, reg_scha63t reg_addr, uint8_t* val)
+{
+    bool ret = false;
+    uint8_t cmd[4];
+    uint8_t bCrc;
+
+    cmd[1] = cmd[2] = 0;
+    cmd[0] = reg_addr << 2;
+    cmd[0] &= 0x7f;
+    cmd[3] = crc8_sae(cmd, 3);
+
+    uint8_t buf[4];
+    switch ( uno_due ) {
+    case SCHA63T_UNO:
+        memcpy(buf, cmd, 4);
+        ret = dev_uno->transfer(buf, 4, buf, 4);
+        memcpy(val, buf, 4);
+        break;
+    case SCHA63T_DUE:
+        memcpy(buf, cmd, 4);
+        ret = dev_due->transfer(buf, 4, buf, 4);
+        memcpy(val, buf, 4);
+        break;
+    default:
+        break;
+    }
+
+    if (ret == true) {
+        bCrc = crc8_sae(val, 3);
+        if ( bCrc != val[3] ) {
```

Comment:
this accesses beyond the end of the val[] array, as callers declare the array as size 3 on the stack

Reply:
Thank you.
I changed it.